### PR TITLE
[TACHYON-542] Evictor Test Framework

### PR DESCRIPTION
**Pull request by uronce-cc**
reopen https://github.com/amplab/tachyon/pull/1033 to master branch.

### CODE REVIEW

**Comment by calvinjia** [comment]
Jenkins, test this please.

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+import com.google.common.io.Files;
+
+import tachyon.Constants;
+import tachyon.worker.block.BlockMetadataManager;
+import tachyon.worker.block.BlockStoreLocation;
+import tachyon.worker.block.meta.StorageDir;
+import tachyon.worker.block.meta.StorageTier;
+
+/**
+ * This class tests general behavior of all types of {@link Evictor} defined in {@link EvictorType}.
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: Can you be  a little more specific here --- e.g., what the general behaviors?

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    return list;
+  }
+
+  public EvictorTest(EvictorType evictorType) {
+    mEvictorType = evictorType;
+  }
+
+  @Before
+  public final void before() throws IOException {
+    mTempFolder = Files.createTempDir();
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: For junit, let 's use to get temp folder so we don't need to clean it up?
```java
@Rule
public TemporaryFolder mTestFolder = new TemporaryFolder();```

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    EvictorTestUtils.cache(USER_ID, BLOCK_ID, cachedBytes, dir, mMeta, mEvictor);
+    Assert.assertTrue(mEvictor.freeSpace(capacity - cachedBytes, EvictorTestUtils.location(dir))
+        .isEmpty());
+  }
+
+  @Test
+  public void noNeedToEvictTest3() throws IOException {
+    // fill in all dirs except for a random directory in a random tier, then request the capacity of
+    // the directory with anyDirInTier
+    StorageDir dirLeft = EvictorTestUtils.randomDir(mMeta);
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: In unit test, it is desired to make everything deterministic and avoid having randomness, so failures could be repeat easily. You can simply pick one dir according to the default manager

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+  }
+
+  @After
+  public final void after() throws IOException {
+    FileUtils.forceDelete(mTempFolder);
+  }
+
+  @Test
+  public void noNeedToEvictTest1() throws IOException {
+    // metadata manager is just created, no cached block in Evictor
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: NIT: I really appreciate the description of the test setting in each test case!! One trick to make the test description even better is to also describe what we expect from the test. For example in this test case, add in the above comment that, we expect the proposed plan is empty. :)

**Comment by apc999** [comment]
Hi @uronce-cc , I really appreciate this patch as it is very important!
I left a few comments--but haven't finished reading all code. I will revisit this PR soon. But meanwhile, feel free to address my comments with followup commits.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java [hunk 5]
```diff
@@ -44,6 +44,16 @@ public EvictionPlan(List<Pair<Long, BlockStoreLocation>> toTransfer, List<Long>
     return mToEvict;
   }
 
+  /**
+   * Whether the plan is empty, an empty plan means it is not null and both toMove and toEvict are
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java]: Actually, I think it is helpful to also javadoc here that, an empty plan indicates no action (move or evict) needs to be taken to meet the requirement.

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+ * requested space can not be satisfied anyway and so on.
+ *
+ * Behavior for a specific type of evictor will be tested in other classes, e.x. tests to ensure
+ * that blocks evicted by LRUEvictor are in the right order should be in LRUEvictorTest.
+ */
+@RunWith(Parameterized.class)
+public class EvictorTest {
+  private static final int USER_ID = 2;
+  private static final long BLOCK_ID = 10;
+  private static final int TEST_TIER = 0;
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: `TEST_TIER_LEVEL` to be more clear? It is pretty bad that we have both concept of level and aliasValue---both in int---to keep backwards compatibility. We plan to fix it in the future. But for now, let's be very specific that the int value is when it refers to a tier, so we can fix all references easier later. Thanks

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+import tachyon.worker.block.BlockMetadataManager;
+import tachyon.worker.block.BlockStoreLocation;
+import tachyon.worker.block.meta.StorageDir;
+import tachyon.worker.block.meta.StorageTier;
+
+/**
+ * This is a parameterized unit test for all types of {@link Evictor} defined in {@link EvictorType}
+ *
+ * It tests general evictor behavior like not evicting any space when space request is already
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: NIT:  rewording suggestion: "It performs sanity check on Evictors regardless of their types, in cases like not evicting any blocks when the required space is already available, proposed eviction ensuring enough space, and returning null eviction plan when the requests can not be achieved."

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    List<Object[]> list = new ArrayList<Object[]>();
+    for (EvictorType type : EvictorType.values()) {
+      if (type != EvictorType.DEFAULT) {
+        list.add(new Object[] {type});
+      }
+    }
+    return list;
+  }
+
+  public EvictorTest(EvictorType evictorType) {
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: I guess constructor should be after members even in junit? Please doublecheck it

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    long totalCapacity = mMetaManager.getAvailableBytes(BlockStoreLocation.anyTier());
+    StorageDir dir = mTestDir;
+    BlockStoreLocation dirLocation = dir.toBlockStoreLocation();
+    long dirCapacity = mMetaManager.getAvailableBytes(dirLocation);
+
+    EvictorTestUtils.cache(USER_ID, BLOCK_ID, dirCapacity, dir, mMetaManager, mEvictor);
+
+    // request space larger than total capacity, no eviction plan should be available
+    Assert.assertNull(mEvictor.freeSpace(totalCapacity + 1, BlockStoreLocation.anyTier()));
+    // request space larger than capacity for the random directory, no eviction plan should be
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: not random any more :)

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTestUtils.java [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+   */
+  // TODO: implement this with shared method in LRUEvictor after the
+  // PR: https://github.com/amplab/tachyon/pull/1079 is merged
+  public static boolean legalCascadingPlan(long bytesToBeAvailable, EvictionPlan plan,
+      BlockMetadataManager metaManager) throws IOException {
+    return true;
+  }
+
+  /**
+   * only when plan is not null and at least one of {@link #legalCascadingPlan},
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTestUtils.java]: only -> Only, to keep consistent with other javadoc

**Comment by uronce-cc** [comment]
@apc999 thanks for the detailed comments including rewording! I've resolved them :)

File: servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java (hunk 8 above)
Comment (uronce-cc) [inline servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java]: yes.

**Comment by calvinjia** [comment]
+1 LGTM

**Pull request merged**